src/rust/lib/data/src/hash_map_tree.rs

```diff
@@ -92,22 +233,68 @@ where K:Eq+Hash {
 
 // === Impls ===
 
-impl<'a,K,T> IntoIterator for &'a HashMapTree<K,T> {
-    type Item     = (&'a K, &'a HashMapTree<K,T>);
-    type IntoIter = std::collections::hash_map::Iter<'a,K,HashMapTree<K,T>>;
-
-    #[inline]
-    fn into_iter(self) -> Self::IntoIter {
-        self.branches.iter()
+impl<K,V,S> Semigroup for HashMapTree<K,V,S>
+    where K : Eq + Hash + Clone,
+          V : Semigroup,
+          S : BuildHasher + Clone {
+    fn concat_mut(&mut self, other:&Self) {
+        self.value.concat_mut(&other.value);
+        self.branches.concat_mut(&other.branches);
     }
 }
 
-impl<'a,K,T> IntoIterator for &'a mut HashMapTree<K,T> {
-    type Item     = (&'a K, &'a mut HashMapTree<K,T>);
-    type IntoIter = std::collections::hash_map::IterMut<'a,K,HashMapTree<K,T>>;
 
-    #[inline]
-    fn into_iter(self) -> Self::IntoIter {
-        self.branches.iter_mut()
-    }
+// === Iterators ===
+
+macro_rules! define_borrow_iterator {
+    ($tp_name:ident $fn_name:ident $($mut:tt)?) => {
+        /// Iterator.
+        pub struct $tp_name<'a,K,V,S> {
+            iters : Vec<std::collections::hash_map::$tp_name<'a,K,HashMapTree<K,V,S>>>,
+            path  : Vec<&'a K>,
+        }
+
+        impl<'a,K,V,S> Iterator for $tp_name<'a,K,V,S> {
+            type Item = (Vec<&'a K>, &'a $($mut)? V);
+            fn next(&mut self) -> Option<Self::Item> {
+                loop {
+                    match self.iters.pop() {
+                        None => break None,
+                        Some(mut iter) => {
+                            match iter.next() {
+                                None => { self.path.pop(); }
+                                Some((sub_key,sub_tree)) => {
+                                    self.iters.push(iter);
+                                    self.iters.push(sub_tree.branches.$fn_name());
+                                    self.path.push(sub_key);
+                                    break Some((self.path.clone(),& $($mut)? sub_tree.value))
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        impl<'a,K,V,S> IntoIterator for &'a $($mut)? HashMapTree<K,V,S> {
+            type Item     = (Vec<&'a K>,&'a $($mut)? V);
+            type IntoIter = $tp_name<'a,K,V,S>;
+
+            #[inline]
+            fn into_iter(self) -> Self::IntoIter {
+                let iters = vec![self.branches.$fn_name()];
+                let path  = default();
+                $tp_name {iters,path}
+            }
+        }
+
+        impl<'a,K,V,S> Debug for $tp_name<'a,K,V,S> {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                write!(f,stringify!($tp_name))
+            }
+        }
+    };
 }
+
+define_borrow_iterator!(Iter iter);
+define_borrow_iterator!(IterMut iter_mut mut);
```

Comment:
Much logic introduced, but is not tested. In fact the whole HashMapTree should be covered with test in this module.

If you've got no time, leave at least TODO